File: android/SwiftKotlination/features/topstories/src/test/kotlin/fr/jhandguy/topstories/model/mocks/TopStoriesManagerMock.kt
package fr.jhandguy.topstories.model.mocks

import fr.jhandguy.network.model.observer.Disposable
import fr.jhandguy.network.model.observer.Observer
import fr.jhandguy.network.model.observer.Result
import fr.jhandguy.topstories.model.TopStories
import fr.jhandguy.topstories.model.TopStoriesManagerInterface

class TopStoriesManagerMock(var result: Result<TopStories>, var observer: Observer<TopStories> = {}) : TopStoriesManagerInterface {
    override suspend fun topStories(observer: Observer<TopStories>): Disposable {
        this.observer = observer
        observer(result)

        return Disposable {}
    }

    override suspend fun fetchStories() = observer(result)
}

File: ios/Modules/TestKit/TestKit/Sources/Mocks/Features/TopStories/Model/TopStoriesManagerMock.swift
import NetworkKit
import StoryKit
import TopStoriesKit

public final class TopStoriesManagerMock {
    public var result: Result<[Story], Error>
    private(set) var observer: Observer<[Story]>

    public init(result: Result<[Story], Error> = .failure(NetworkError.invalidResponse)) {
        self.result = result
        observer = { _ in }
    }
}

extension TopStoriesManagerMock: TopStoriesManagerProtocol {
    @discardableResult
    public func stories(_ observer: @escaping Observer<[Story]>) -> Disposable {
        self.observer = observer
        observer(result)

        return Disposable {}
    }

    public func fetchStories() {
        stories(observer)
    }
}
